src/Core/Types/Utilities/TypeConversion.OptionalTypes.cs

```diff
@@ -0,0 +1,48 @@
+using System;
+
+namespace HotChocolate.Utilities
+{
+    public partial class TypeConversion
+    {
+        private bool TryCreateOptionalConverter(
+            Type from, Type to, out ChangeType converter)
+        {
+            Type innerFrom = GetUnderlyingOptionalType(from);
+            Type innerTo = GetUnderlyingOptionalType(to);
+
+            if ((innerFrom != from || innerTo != to)
+                && TryGetOrCreateConverter(innerFrom, innerTo, out converter))
+            {
+                ChangeType innerConverter = converter;
+                converter = source => GenericOptionalConverter((IOptional)source, innerConverter);
```

Comment:
what happens if the source is not optional but the target type?

I think I will rewrite this part and compile a converter expression for this.

We have the following cases here:

- from is not an `Optional<T>` but to is an `Optional<T>`
- from is an `Optional<T>` but to is not an `Optional<T>`
- from and to are `Optional<T>`

Reply:
Yea i considered that scenario, but made a best guess given the existing code. And it didn't seem like the assumption broke anything, but you know best :)

Reply:
With nullable, it is different since, in the end, it does not really matter since if there is a value you always have the int not a different type. But in this case, we would run into conversion exceptions I think. But I think I can work this one out tomorrow morning.

Reply:
Can I do anything to help get this over the finish line? As mentioned, this is pretty urgent for us, since it blocks progress on quite a few things for us.

The only thing missing should be this: `from is not an Optional<T> but to is an Optional<T>`, the other two scenarious should already be covered as i see it? Should I add something that can handle this scenario too? Is there really a benefit to using an expression here? It adds the converter for re-use anyway.

Reply:
Not at this moment... I have started reworking this part on 11 and will port it back once ready.